lib/mocha.js

```diff
@@ -866,3 +872,46 @@ Mocha.prototype.run = function(fn) {
 
   return runner.run(done);
 };
+
+/**
+ * Assigns hooks to the root suite
+ * @param {MochaRootHookObject} [hooks] - Hooks to assign to root suite
+ * @chainable
+ */
+Mocha.prototype.rootHooks = function rootHooks(hooks) {
+  if (utils.type(hooks) === 'object') {
+    var beforeAll = [].concat(hooks.beforeAll || []);
+    var beforeEach = [].concat(hooks.beforeEach || []);
+    var afterAll = [].concat(hooks.afterAll || []);
+    var afterEach = [].concat(hooks.afterEach || []);
+    var rootSuite = this.suite;
+    beforeAll.forEach(function(hook) {
+      rootSuite.beforeAll(hook);
+    });
+    beforeEach.forEach(function(hook) {
+      rootSuite.beforeEach(hook);
+    });
+    afterAll.forEach(function(hook) {
+      rootSuite.afterAll(hook);
+    });
+    afterEach.forEach(function(hook) {
+      rootSuite.afterEach(hook);
+    });
+  }
+  return this;
+};
+
+/**
+ * An alternative way to define root hooks that works with parallel runs.
+ * @typedef {Object} MochaRootHookObject
```

Comment:
Do we need this typedef for typescript? Because this file is part of our bundled browser mocha.js.

Reply:
I can move it elsewhere.  It's not intended to be in the public API at this point anyway.  VSCode's TS language server understands it (somewhat).  I just did this instead of doing it inline elsewhere, as describing more than the most trivial values in a `@param` is awkward.

Given that this functionality is intended to be specific to Node.js, maybe I should move it out?  That would be poor encapsulation, though.  I really dislike the points in the codebase where object _A_ is mucking with the internals of object _B_, and that's exactly what would have to happen if `Mocha#rootHooks()` were moved elsewhere; we'd have a `Mocha` instance, then some other method would need to reach into `Mocha#suite` and call methods there.

While this is intended to be used with Node.js, it _could_ be used in the browser.  There's no reason why a browser _can't_ call `Mocha#rootHooks()`, but the consumer would need to pass an object adhering to the  `MochaRootHookObject` shape, which is awkward.

Another way may be to do away with `MochaRootHookObject` entirely, and instead expose four new chainable methods: `Mocha#rootBeforeAll`, `Mocha#rootBeforeEach`, `Mocha#rootAfterAll`, and `Mocha#rootAfterEach`. Each would accept a single hook function.  This would probably be a more ergonomic API for a consumer (which I had not originally considered).

What do you think?

Reply:
I mean, there's no reason why the browser _couldn't_ consume `rootHooks()`.  I think it's probably the right place for it for now.